### PR TITLE
Introduce master log compression.

### PR DESCRIPTION
**Pull request by bastien31**
Log in the db is stored as chunck of text.
The implementation expected a compress operation to be done at the end
of the step. But it will be costly in time/db operations with db
fragmentation risk.

This patch suggest to do live compression when chunck size to insert
is smaller once compressed.
The compress operation will "just" have then to optimize/regroup
all contiguous text chunk, compress them and replace them in the db.
Compression algorithm is configurable in the master cfg.

Signed-off-by: Sebastien Fusilier <sebastien.fusilier@intel.com>

### CODE REVIEW

**Comment by ssavrim** [comment]
:+1:

File: master/buildbot/config.py [hunk 1]
```diff
@@ -297,8 +297,13 @@ def copy_str_param(name, alt_key=None):
...
-            if logCompressionMethod not in ('bz2', 'gz'):
-                error("c['logCompressionMethod'] must be 'bz2' or 'gz'")
+            if logCompressionMethod not in ('raw', 'bz2', 'gz', 'lz4'):
+                error("c['logCompressionMethod'] must be 'raw', 'bz2', 'gz' or 'lz4'")
+            elif logCompressionMethod == "lz4":
+                try:
+                    import lz4
+                except ImportError:
+                    error("To set c['logCompressionMethod'] to 'lz4' you must install the lz4 library ('pip install lz4')")
             self.logCompressionMethod = logCompressionMethod
```
Comment (tardyp) [inline master/buildbot/config.py]: else:
       self.logCompressionMethod = 'raw'

File: master/buildbot/db/logs.py [hunk 2]
```diff
@@ -119,20 +149,32 @@ def thd(conn):
...
+
+                # Set the default compressed mode to "raw" id
+                compressed_mode = self.COMPRESSION_MODE["raw"]["id"]
+                # Do we have to compress the chunk?
+                if self.master.config.logCompressionMethod != "raw":
+                    start = time.clock()
+                    compressed_chunk = self.COMPRESSION_MODE[self.master.config.logCompressionMethod]["dumps"](chunk)
+                    # Is it useful to compress the chunk?
+                    if len(chunk) > len(compressed_chunk):
+                        compressed_mode = self.COMPRESSION_MODE[self.master.config.logCompressionMethod]["id"]
```
Comment (tardyp) [inline master/buildbot/db/logs.py]: please factorize
candidate_compressed_mode = self.COMPRESSION_MODE[self.master.config.logCompressionMethod]

File: master/buildbot/db/logs.py [hunk 3]
```diff
@@ -155,7 +197,7 @@ def _splitBigChunk(self, content, logid):
         truncline = content[:self.MAX_CHUNK_SIZE]
         while truncline:
             try:
-                truncline.decode('utf-8')
+                truncline.decode(self.master.config.logEncoding)
```
Comment (tardyp) [inline master/buildbot/db/logs.py]: no, we should remove logEncoding, and put everything in utf-8.

All data in db shall be utf-8

File: master/buildbot/config.py (hunk 1 above)
Comment (bastien31) [inline master/buildbot/config.py]: And so replace error("...") to log.msg => don't you think that the user may not see the error and continue running the master thinking that the log is compressed?
Stopping is more radical but more "informative".

Comment (tardyp) [inline master/buildbot/config.py]: I meant the else for the if 'logCompressionMethod' not in config_dict:

File: master/buildbot/db/logs.py (hunk 3 above)
Comment (bastien31) [inline master/buildbot/db/logs.py]: Ok I restore back the hard-coded 'utf-8', removing logEncoding param will be done in another patch.

**Comment by tardyp** [comment]
the unit tests are broken

File: master/buildbot/db/logs.py [hunk 4]
```diff
@@ -68,8 +95,11 @@ def thd(conn):
             q = q.order_by(tbl.c.first_line)
             rv = []
             for row in conn.execute(q):
-                assert not row.compressed, "compressed rows not supported yet"
-                content = row.content.decode('utf-8')
+                # Retrieve associated "reader" and extract the data
+                data = [y["read"] for y in self.COMPRESSION_MODE.itervalues() if y["id"] == row.compressed][0](
```
Comment (tardyp) [inline master/buildbot/db/logs.py]: would be better if we have a dict to directly map id to read value.

File: master/buildbot/db/logs.py [hunk 5]
```diff
@@ -122,17 +152,29 @@ def thd(conn):
...
+
+                # Set the default compressed mode to "raw" id
+                compressed_mode = self.COMPRESSION_MODE["raw"]["id"]
+                # Do we have to compress the chunk?
+                if self.master.config.logCompressionMethod != "raw":
+                    compressed_mode = self.COMPRESSION_MODE[self.master.config.logCompressionMethod]
+                    compressed_chunk = compressed_mode["dumps"](chunk)
+                    # Is it useful to compress the chunk?
+                    if len(chunk) > len(compressed_chunk):
+                        compressed_mode = compressed_mode["id"]
```
Comment (tardyp) [inline master/buildbot/db/logs.py]: looks like the else of this check is not tested, as compressed_mode has to be reset to 0

**Comment by bastien31** [comment]
Need some more unit test. But at least it should be "merge-able".

**Comment by sa2ajj** [comment]
```
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/internet/defer.py", line 1037, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/failure.py", line 382, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_db_logs.py", line 281, in test_addLogLines_big_chunk
    (yield self.db.logs.appendLog(201, u'abc\n' * 20000)),  # 80k
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/threadpool.py", line 207, in _worker
    result = context.call(ctx, function, *args, **kwargs)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/pool.py", line 191, in __thd
    rv = callable(arg, *args, **kwargs)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/logs.py", line 170, in thd
    log.msg("Store {} instead of {}".format(len(chunk), raw_len))
exceptions.ValueError: zero length field name in format

buildbot.test.unit.test_db_logs.TestRealDB.test_addLogLines_big_chunk
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/internet/defer.py", line 1037, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/failure.py", line 382, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_db_logs.py", line 291, in test_addLogLines_big_chunk_big_lines
    self.assertEqual((yield self.db.logs.appendLog(201, line * 3)),
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/threadpool.py", line 207, in _worker
    result = context.call(ctx, function, *args, **kwargs)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/pool.py", line 191, in __thd
    rv = callable(arg, *args, **kwargs)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/logs.py", line 170, in thd
    log.msg("Store {} instead of {}".format(len(chunk), raw_len))
exceptions.ValueError: zero length field name in format

buildbot.test.unit.test_db_logs.TestRealDB.test_addLogLines_big_chunk_big_lines
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/internet/defer.py", line 1037, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/failure.py", line 382, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_db_logs.py", line 325, in test_addLogLines_huge_lines
    yield self.db.logs.appendLog(201, line * 3)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/threadpool.py", line 207, in _worker
    result = context.call(ctx, function, *args, **kwargs)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/pool.py", line 191, in __thd
    rv = callable(arg, *args, **kwargs)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/logs.py", line 170, in thd
    log.msg("Store {} instead of {}".format(len(chunk), raw_len))
exceptions.ValueError: zero length field name in format

buildbot.test.unit.test_db_logs.TestRealDB.test_addLogLines_huge_lines
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/internet/defer.py", line 1037, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/failure.py", line 382, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_db_logs.py", line 419, in test_bz2_compress_big_chunk
    (yield self.db.logs.appendLog(201, line + '\n')),
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/threadpool.py", line 207, in _worker
    result = context.call(ctx, function, *args, **kwargs)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/pool.py", line 191, in __thd
    rv = callable(arg, *args, **kwargs)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/logs.py", line 170, in thd
    log.msg("Store {} instead of {}".format(len(chunk), raw_len))
exceptions.ValueError: zero length field name in format

buildbot.test.unit.test_db_logs.TestRealDB.test_bz2_compress_big_chunk
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/internet/defer.py", line 1037, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/failure.py", line 382, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_db_logs.py", line 394, in test_gz_compress_big_chunk
    (yield self.db.logs.appendLog(201, line + '\n')),
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/threadpool.py", line 207, in _worker
    result = context.call(ctx, function, *args, **kwargs)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/pool.py", line 191, in __thd
    rv = callable(arg, *args, **kwargs)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/logs.py", line 170, in thd
    log.msg("Store {} instead of {}".format(len(chunk), raw_len))
exceptions.ValueError: zero length field name in format

buildbot.test.unit.test_db_logs.TestRealDB.test_gz_compress_big_chunk
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/internet/defer.py", line 1037, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/failure.py", line 382, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_db_logs.py", line 444, in test_lz4_compress_big_chunk
    (yield self.db.logs.appendLog(201, line + '\n')),
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/threadpool.py", line 207, in _worker
    result = context.call(ctx, function, *args, **kwargs)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/pool.py", line 191, in __thd
    rv = callable(arg, *args, **kwargs)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/db/logs.py", line 170, in thd
    log.msg("Store {} instead of {}".format(len(chunk), raw_len))
exceptions.ValueError: zero length field name in format

buildbot.test.unit.test_db_logs.TestRealDB.test_lz4_compress_big_chunk
```

**Comment by tardyp** [comment]
FWYI, I have made some tests on a copy of the nine.buildbot.net db.

The current sqlite db is 6.1GB

I first simulated seb's method of on the fly compression of the chunks (using bzip algorithm)
I created a version of the db, which does not concatenate the chunks, and just compress them
the cleanedup is 721MB, which is a divide the size by 8.5.
I created another version of the db, which concatenate all chunks of each log, and compress them altogether.
the cleanedup is 458MB, which is a divide the size by 13.4.

I'll share my "buildbot cleanupdb" command line after seb's lands.

File: master/buildbot/test/unit/test_db_logs.py [hunk 6]
```diff
@@ -337,6 +337,128 @@ def test_splitBigChunk_unicode_misalignment(self):
...
+        self.assertEqual(newRow, {
+            'logid': 201,
+            'first_line': 7,
+            'last_line': 7,
+            'content': bz2.compress(line, 9),
+            'compressed': 2})
+
+    @defer.inlineCallbacks
+    def test_lz4_compress_big_chunk(self):
+        import lz4
```
Comment (tardyp) [inline master/buildbot/test/unit/test_db_logs.py]: need to skip if lz4 is not present. See txgithub skips

File: master/buildbot/config.py [hunk 7]
```diff
@@ -295,11 +295,16 @@ def copy_str_param(name, alt_key=None):
...
+        self.logCompressionMethod = config_dict.get('logCompressionMethod', 'bz2')
+        if self.logCompressionMethod not in ('raw', 'bz2', 'gz', 'lz4'):
+            error("c['logCompressionMethod'] must be 'raw', 'bz2', 'gz' or 'lz4'")
+
+        if self.logCompressionMethod == "lz4":
+            try:
+                import lz4
+            except ImportError:
+                error("To set c['logCompressionMethod'] to 'lz4' you must install the lz4 library"
+                      "('pip install lz4'). Compression mode is set to 'raw'")
```
Comment (seankelly) [inline master/buildbot/config.py]: Message says the compression mode is set to raw, but an error is raised instead.

**Comment by seankelly** [comment]
The default compression method is bz2 but in the compression table, gzip has the best compression ratio and is almost twice as fast as bz2. I think gzip should be the default method because of its overall simplicity and speed.

**Comment by sa2ajj** [comment]
/me agrees w/ @seankelly: gzip sounds like a better default _and_ it does not require extra dependencies.

**Comment by tardyp** [comment]
agrees also, for the nine.bb.net database I made this night benchmark on the compression method:
I recompress the same 6GB db with real data from nine.bb.net:

437M	state.sqlite.bz2
442M	state.sqlite.gz
729M	state.sqlite.lz4

I'll make some throughput tests for completeness

File: master/buildbot/config.py (hunk 7 above)
Comment (sa2ajj) [inline master/buildbot/config.py]: I think, showing an error is a better approach than falling back to some other mode.

Comment (seankelly) [inline master/buildbot/config.py]: Agreed, but the error message implies it is falling back to something else. Except it's an error so it can't fall back to anything.

Comment (bastien31) [inline master/buildbot/config.py]: Yes, I did implement a fallback but it was not a good idea => I did remove it but the error message was not updated. I do the fix.

**Comment by bastien31** [comment]
Coveralls.io is stuck? Else the pull request shall be ready. Waiting for reviewers...

**Comment by tardyp** [comment]
:+1:

**Comment by sa2ajj** [comment]
LGTM

**Comment by bastien31** [comment]
Any ETA for this LGTM?

**Comment by tardyp** [comment]
as there are at least 2 approval, I merge this PR

**Comment by bastien31** [comment]
Thx!
